### PR TITLE
Redesign Semantic Tree Compilation Algorithm

### PR DESCRIPTION
**Pull request by goderbauer**
* removes the second pass
* allows a parent to take annotations from multiple children (required for multiple text nodes to be merged)
* nodes can require that their children cannot annotate them at all (explicitChildNodes)
* RenderObjects describe what they want to add to the semantics tree in a SemanticsConfiguration object
* do not generate semantics update if nothing has changed (even after calls to markNeedsSemanticsUpdate)
* fix a bug where clips were not applied past semantic boundaries during updates
* simplifications: remove unused logic, etc.
* some more tests

### CODE REVIEW

File: examples/flutter_gallery/flutter_gallery.iml [hunk 1]
```diff
@@ -10,5 +10,6 @@
     <orderEntry type="sourceFolder" forTests="false" />
     <orderEntry type="library" name="Dart Packages" level="project" />
     <orderEntry type="library" name="Dart SDK" level="project" />
+    <orderEntry type="library" name="Flutter Plugins" level="project" />
   </component>
-</module>
+</module>
```
Comment (Hixie) [inline examples/flutter_gallery/flutter_gallery.iml]: remove from this PR

File: packages/flutter/lib/src/material/tabs.dart [hunk 2]
```diff
@@ -742,22 +742,20 @@ class _TabBarState extends State<TabBar> {
...
+      wrappedTabs[index] = new InkWell(
+        onTap: () { _handleTap(index); },
+        child: new Padding(
+          padding: new EdgeInsets.only(bottom: widget.indicatorWeight),
+          child: new Stack(
+            children: <Widget>[
+              wrappedTabs[index],
+              new Semantics(
+                selected: index == _currentIndex,
+                // TODO(goderbauer): I10N-ify
```
Comment (Hixie) [inline packages/flutter/lib/src/material/tabs.dart]: cc @HansMuller

File: packages/flutter/lib/src/rendering/proxy_box.dart [hunk 3]
```diff
@@ -2987,16 +2987,46 @@ class RenderSemanticsGestureHandler extends RenderProxyBox implements SemanticsA
...
 
   @override
-  SemanticsAnnotator get semanticsAnnotator => isSemanticBoundary ? _annotate : null;
+  void describeSemanticsConfiguration(SemanticsConfiguration config) {
+    super.describeSemanticsConfiguration(config);
+
+    config.isSemanticBoundary = _hasHandlers;
+
+    // TODO(goderbauer): this needs to be set even when there is only potential
+    //    for this to become a scroll view.
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/proxy_box.dart]: any particular reason not to do this now?

File: packages/flutter/lib/src/rendering/proxy_box.dart [hunk 4]
```diff
@@ -3163,12 +3151,9 @@ class RenderSemanticsAnnotations extends RenderProxyBox {
   /// node in the semantics tree. Otherwise, the semantics will be
   /// merged with the semantics of any ancestors.
   ///
-  /// The 'container' flag is implicitly set to true on the immediate
-  /// semantics-providing descendants of a node where multiple
-  /// children have semantics or have descendants providing semantics.
-  /// In other words, the semantics of siblings are not merged. To
-  /// merge the semantics of an entire subtree, including siblings,
-  /// you can use a [RenderMergeSemantics].
+  /// Whether decedents of this RenderObject can add their semantic information
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/proxy_box.dart]: "decedents" is presumably a typo

Comment (Hixie) [inline packages/flutter/lib/src/rendering/proxy_box.dart]: `[RenderObject]`

File: packages/flutter/lib/src/rendering/proxy_box.dart [hunk 5]
```diff
@@ -3179,6 +3164,28 @@ class RenderSemanticsAnnotations extends RenderProxyBox {
     markNeedsSemanticsUpdate();
   }
 
+  /// Whether decedents of this RenderObject are allowed to add semantic
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/proxy_box.dart]: ditto ditto

File: packages/flutter/lib/src/rendering/proxy_box.dart [hunk 6]
```diff
@@ -3179,6 +3164,28 @@ class RenderSemanticsAnnotations extends RenderProxyBox {
...
+  /// information to the [SemanticsNode] annotated by this widget.
+  ///
+  /// When set to false decedents are allowed to annotate [SemanticNode]s of
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/proxy_box.dart]: ditto

File: packages/flutter/lib/src/rendering/proxy_box.dart [hunk 7]
```diff
@@ -3179,6 +3164,28 @@ class RenderSemanticsAnnotations extends RenderProxyBox {
...
+  /// their parent with the semantic information they want to contribute to the
+  /// semantic tree.
+  /// When set to true the only way for decedents to contribute semantic
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/proxy_box.dart]: ditto

Comment (Hixie) [inline packages/flutter/lib/src/rendering/proxy_box.dart]: if you were hoping for a paragraph break here you won't get one (if you weren't, it's fine)

File: packages/flutter/lib/src/rendering/object.dart [hunk 8]
```diff
@@ -2528,44 +2115,26 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
...
   }
 
-  /// Whether this [RenderObject] introduces a new box for accessibility purposes.
-  ///
-  /// See also:
-  ///
-  ///  * [semanticsAnnotator], which fills in the [SemanticsNode] implied by
-  ///    setting [isSemanticBoundary] to true.
-  bool get isSemanticBoundary => false;
+  void describeSemanticsConfiguration(SemanticsConfiguration config) {
```
Comment (Hixie) [inline packages/flutter/lib/src/rendering/object.dart]: @protected
@mustCallSuper ?

**Comment by Hixie** [comment]
LGTM

File: packages/flutter/lib/src/rendering/proxy_box.dart (hunk 3 above)
Comment (goderbauer) [reply packages/flutter/lib/src/rendering/proxy_box.dart]: Let's do this in a follow-up. This requires some plumbing and this PR is already pretty big.

**Pull request merged**